jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/content/ContentSourcePublisherTest.java

```diff
@@ -0,0 +1,249 @@
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.io.content;
+
+import java.nio.ByteBuffer;
+import java.util.Random;
+import java.util.concurrent.Flow;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.eclipse.jetty.io.Content;
+import org.reactivestreams.tck.TestEnvironment;
+import org.reactivestreams.tck.flow.FlowPublisherVerification;
+
+public final class ContentSourcePublisherTest extends FlowPublisherVerification<Content.Chunk>
+{
+    public ContentSourcePublisherTest()
+    {
+        super(new TestEnvironment());
+    }
+
+    @Override
+    public Flow.Publisher<Content.Chunk> createFlowPublisher(long elements)
+    {
+        Content.Source source = new SyntheticContentSource(elements);
+        return new ContentSourcePublisher(source);
+    }
+
+    @Override
+    public Flow.Publisher<Content.Chunk> createFailedFlowPublisher()
+    {
+        Content.Source source = new SyntheticContentSource(0);
+        Flow.Publisher<Content.Chunk> publisher = new ContentSourcePublisher(source);
+        // Simulate exhausted Content.Source
+        publisher.subscribe(new Flow.Subscriber<>()
+        {
+            @Override
+            public void onSubscribe(Flow.Subscription subscription)
+            {
+                subscription.cancel();
+            }
+
+            @Override
+            public void onNext(Content.Chunk item)
+            {
+            }
+
+            @Override
+            public void onError(Throwable throwable)
+            {
+            }
+
+            @Override
+            public void onComplete()
+            {
+            }
+        });
+        return publisher;
+    }
+
+    private static final class SyntheticContentSource implements Content.Source
+    {
+        private final AtomicReference<State> state;
+        private final long contentSize;
+
+        public SyntheticContentSource(long chunksToRead)
+        {
+            this.state = new AtomicReference<>(new State.Reading(chunksToRead));
+            this.contentSize = State.Reading.chunkSize * Math.max(chunksToRead, 0);
+        }
+
+        @Override
+        public long getLength()
+        {
+            return contentSize;
+        }
+
+        @Override
+        public Content.Chunk read()
+        {
+            return state.getAndUpdate(before -> before.read()).chunk();
+        }
+
+        @Override
+        public void demand(Runnable demandCallback)
+        {
+            // TODO: recursive stack overflow
```

Comment:
I'm not sure if it necessary to handle this, but according to the contract demand must be safely reentrant

Reply:
Typically we would implement this with a `SerialInvoker`, but for a test, it is probably not necessary.